experimental/attestation_common/src/crypto.rs

```diff
@@ -0,0 +1,123 @@
+        Self(Some(nonce))
+    }
+}
+
+impl aead::NonceSequence for OneNonceSequence {
+    fn advance(&mut self) -> Result<aead::Nonce, ring::error::Unspecified> {
+        self.0.take().ok_or(ring::error::Unspecified)
+    }
+}
+
+pub struct AeadEncryptor {
+    // Key used for encrypting data.
+    sealing_key: aead::SealingKey<OneNonceSequence>,
+    // Key used for decrypting data.
+    opening_key: aead::OpeningKey<OneNonceSequence>,
+}
+
+impl AeadEncryptor {
+    pub fn new(key: &[u8]) -> Self {
+        let unbound_sealing_key = aead::UnboundKey::new(&aead::AES_256_GCM, &key).unwrap();
+        let sealing_key = ring::aead::SealingKey::new(unbound_sealing_key, OneNonceSequence::new());
+
+        let unbound_opening_key = aead::UnboundKey::new(&aead::AES_256_GCM, &key).unwrap();
+        let opening_key = ring::aead::OpeningKey::new(unbound_opening_key, OneNonceSequence::new());
+
+        Self {
+            sealing_key,
+            opening_key,
+        }
+    }
+
+    pub fn encrypt(&mut self, data: &[u8]) -> anyhow::Result<Vec<u8>> {
+        let mut encrypted_data = data.to_vec();
+        self.sealing_key
+            .seal_in_place_append_tag(
+                // Additional authenticated data is not required.
+                aead::Aad::from(vec![]),
+                &mut encrypted_data,
+            )
+            .map_err(|error| anyhow!("Couldn't encrypt data: {:?}", error))?;
+        Ok(encrypted_data)
+    }
+
+    pub fn decrypt(&mut self, data: &[u8]) -> anyhow::Result<Vec<u8>> {
+        let mut decrypted_data = data.to_vec();
+        let decrypted_data = self
+            .opening_key
+            .open_in_place(
+                // Additional authenticated data is not required.
+                aead::Aad::from(vec![]),
+                &mut decrypted_data,
+            )
+            .map_err(|error| anyhow!("Couldn't decrypt data: {:?}", error))?;
+        Ok(decrypted_data.to_vec())
+    }
+}
+
+pub struct DiffieHellmanNegotiator {
+    private_key: agreement::EphemeralPrivateKey,
+}
+
+impl DiffieHellmanNegotiator {
+    pub fn create() -> anyhow::Result<Self> {
+        let rng = rand::SystemRandom::new();
+        let private_key = agreement::EphemeralPrivateKey::generate(&agreement::X25519, &rng)
+            .map_err(|error| anyhow!("Couldn't generate private key: {:?}", error))?;
+        Ok(Self { private_key })
+    }
+
+    pub fn public_key(&self) -> anyhow::Result<Vec<u8>> {
+        let public_key = self
+            .private_key
+            .compute_public_key()
+            .map_err(|error| anyhow!("Couldn't get public key: {:?}", error))?;
+        Ok(public_key.as_ref().to_vec())
+    }
+
+    pub fn derive_session_key(self, peer_public_key: &[u8]) -> anyhow::Result<Vec<u8>> {
+        let peer_public_key =
+            agreement::UnparsedPublicKey::new(&agreement::X25519, peer_public_key);
+        let session_key = agreement::agree_ephemeral(
```

Comment:
Crypto best practice is to never use keys for more than one purpose: which in handshake protocols generally translates to: deriving single-purpose keys from the ephemeral shared secret.
Furthermore,  these keys can/should be bound to the handshake run by using the transcript as HKDF info (as is done in TLS, ALTS/LOAS, EKEP and our simplification of it too).

Reply:
Does this mean that we need to generate a session key not just from `KDF(key_material, server_key, client_key)`, but also from some purpose string?

Created an issue: https://github.com/project-oak/oak/issues/2100

Reply:
Or `KDF(key_material, server_key, client_key)` is enough, because we generate new ephemeral keys for every streaming gRPC request?

Reply:
Ideally KDF(key_material, info=handshake_transcript) where handshake_transcript will include the server and client ephemeral keys.
We actually maintain separate client and server encryption keys (both client and server have both keys, the client key is used by the client to encrypt and server to decrypt, and vice versa). We append the type of the encryption key to the transcript_hash too. This is to bind the keys to the direction of data flow (i.e. as mentioned in another comment, to disallow the untrusted network to loopback responses as requests).